File: native/app/group_outbox_test.go
package app

import (
	"bytes"
	"context"
	"errors"
	"strings"
	"testing"
	"time"

	"github.com/JohnnyPBelo/relayloom/native/core"
	"github.com/JohnnyPBelo/relayloom/native/groupauthority"
	"github.com/JohnnyPBelo/relayloom/native/groupledger"
	"github.com/JohnnyPBelo/relayloom/native/groups"
	"github.com/JohnnyPBelo/relayloom/native/groupstore"
	"github.com/JohnnyPBelo/relayloom/native/transport"
)

// Fixtures install a valid signed/admitted message and intent. The scope here
// is runtime stop/recovery, not dynamic sending or P2P control-carrier delivery.
func groupOutboxFixture(t *testing.T) (*Node, groupauthority.View, OutboxRecord) {
	t.Helper()
	n, identity, group, _ := groupContentFixture(t)
	reader, err := core.CreateIdentity("Reader")
	if err != nil {
		t.Fatal(err)
	}
	var snapshot *groups.GroupSnapshot
	err = n.privateDatabase.Update(func(tx *groupstore.Tx) error {
		_, err := groupledger.Run(tx, identity, func(l *groupledger.Ledger, g *groupauthority.Registry) error {
			anchor, err := g.Anchor(group.ID)
			if err != nil {
				return err
			}
			invitation, err := groups.CreateInvitation(identity, anchor, *group.Head, reader.Public)
			if err != nil {
				return err
			}
			consent, err := groups.AcceptInvitation(reader, anchor, *group.Head, invitation)
			if err != nil {
				return err
			}
			_, err = g.Commit(operationFor(501), group.ID, group.Head.ID, groups.Update{Title: "Outbox", Members: []core.PublicIdentity{identity.Public, reader.Public}, Joins: []groups.GroupConsent{consent}})
			if err != nil {
				return err
			}
			group, err = g.State(group.ID)
			if err != nil {
				return err
			}
			snapshot, err = g.PrivateState(group.ID, group.Head.ID)
			return err
		})
		return err
	})
	if err != nil {
		t.Fatal(err)
	}
	b := contentMessage(t, identity, group, snapshot, group.Head.ID)
	if _, err := n.Store.PutReserved(b, false); err != nil {
		t.Fatal(err)
	}
	apply(t, n, "state", nil)
	encoded, _ := core.Canonical(b)
	stopped := false
	entry := OutboxRecord{OperationID: operationFor(502), Fingerprint: core.Hash([]byte("fixture intent")), ID: b.Manifest.ID,
		Author: identity.Public.ID, Conversation: group.ID, Preview: "Commit before display", Created: b.Manifest.Created,
		Expires: b.Manifest.Expires, Priority: "normal", Bytes: int64(len(encoded)), Phase: "ready", Confirmations: map[string]Confirmation{reader.Public.ID: {}}, GroupEpoch: group.Head.ID, GroupStopped: &stopped}
	next, err := copyPrivate(n.private, identity.Public.ID)
	if err != nil {
		t.Fatal(err)
	}
	next.Outbox[entry.OperationID] = entry
	if err := n.persistPrivateLocked(next); err != nil {
		t.Fatal(err)
	}
	apply(t, n, "state", nil)
	return n, group, entry
}

func closeOutboxGroup(n *Node, group groupauthority.View) (any, error) {
	return n.Handle("group-command", map[string]any{"action": "close", "operationId": operationFor(503), "groupId": group.ID, "expected": group.Head.ID})
}

func TestGroupOutboxMissingSnapshotPausesWithoutLosingAvailability(t *testing.T) {
	n, group, entry := groupOutboxFixture(t)
	snapshot := apply(t, n, "group-command", map[string]any{"action": "private-state", "groupId": group.ID, "epochId": group.Head.ID}).(map[string]any)["snapshot"].(*groups.GroupSnapshot)
	anchor := apply(t, n, "group-command", map[string]any{"action": "proofs", "groupId": group.ID, "from": 0}).(map[string]any)["anchor"].(groups.GroupAnchor)
	successor, err := groups.CreateSuccessor(*n.identity, anchor, *group.Head, *snapshot, groups.Update{Title: "Awaiting proof", Members: snapshot.Members, Joins: []groups.GroupConsent{}})
	if err != nil {
		t.Fatal(err)
	}
	apply(t, n, "group-command", map[string]any{"action": "headers", "groupId": group.ID, "headers": []groups.GroupEpoch{successor.Epoch}})
	item := outboxFor(t, n, entry.OperationID)
	if item["status"] != "pending" || item["retained"] != true || n.private.Outbox[entry.OperationID].Phase != "ready" || n.groupRetries[entry.OperationID].Allowed || n.groupRetries[entry.OperationID].Stop != nil {
		t.Fatal("missing proof became lost content or retry permission")
	}
	apply(t, n, "group-command", map[string]any{"action": "snapshot", "groupId": group.ID, "epochId": successor.Epoch.ID, "snapshot": successor.Snapshot})
	outboxFor(t, n, entry.OperationID)
	if !n.groupRetries[entry.OperationID].Allowed {
		t.Fatal("valid successor did not restore retry permission")
	}
	if err := n.privateDatabase.Update(func(tx *groupstore.Tx) error {
		return tx.Delete("snapshot:" + group.ID + ":" + group.Head.Body.SnapshotHash)
	}); err != nil {
		t.Fatal(err)
	}
	outboxFor(t, n, entry.OperationID)
	if len(n.groupHolds) != 1 || n.private.Outbox[entry.OperationID].Phase != "ready" {
		t.Fatal("missing historical proof lost availability or quarantine")
	}
	apply(t, n, "group-command", map[string]any{"action": "snapshot", "groupId": group.ID, "epochId": group.Head.ID, "snapshot": snapshot})
	outboxFor(t, n, entry.OperationID)
	if !n.groupRetries[entry.OperationID].Allowed || len(n.groupHolds) != 0 || len(n.Store.Reservations()) != 1 {
		t.Fatal("proof recovery retained obsolete hold or lost pending reserve")
	}
}

func TestGroupOutboxMinimalStopAndManualPinSurviveReopen(t *testing.T) {
	n, group, entry := groupOutboxFixture(t)
	if outboxFor(t, n, entry.OperationID)["status"] != "pending" || !equalIDs(n.Store.Reservations(), []string{entry.ID}) {
		t.Fatal("fixture has no pending reserved message")
	}
	if err := n.pinOutboxLocked(entry.ID, true); err != nil {
		t.Fatal(err)
	}
	before, err := n.privateDatabase.Read()
	if err != nil {
		t.Fatal(err)
	}
	if _, err := closeOutboxGroup(n, group); err != nil {
		t.Fatal(err)
	}
	after, err := n.privateDatabase.Read()
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.Equal(before.Bytes, after.Bytes) || before.Digest != after.Digest {
		t.Fatal("minimal stop rewrote private document")
	}
	if outboxFor(t, n, entry.OperationID)["status"] != "superseded" || len(n.Store.Reservations()) != 0 || !n.Store.IsPinned(entry.ID) {
		t.Fatal("stop lost its projection or manual pin")
	}
	stop := n.groupRetries[entry.OperationID].Stop
	apply(t, n, "lock", nil)
	apply(t, n, "unlock", map[string]any{"password": password})
	if outboxFor(t, n, entry.OperationID)["status"] != "superseded" || n.groupRetries[entry.OperationID].Stop.Observed != stop.Observed {
		t.Fatal("reopen lost immutable stop")
	}
	apply(t, n, "outbox-retry", map[string]any{"operationId": entry.OperationID})
	if n.private.Outbox[entry.OperationID].Attempts != 0 {
		t.Fatal("stopped retry attempted transmission")
	}
}

func TestGroupOutboxStopAndAuthorityShareFailureBoundary(t *testing.T) {
	for _, committed := range []bool{false, true} {
		name := "rollback"
		if committed {
			name = "lost-commit-response"
		}
		t.Run(name, func(t *testing.T) {
			n, group, entry := groupOutboxFixture(t)
			db := n.privateDatabase
			injected := errors.New("fixture commit failure")
			n.updateGroupState = func(callback func(*groupstore.Tx) error) error {
				if committed {
					if err := db.Update(callback); err != nil {
						return err
					}
					return injected
				}
				return db.Update(func(tx *groupstore.Tx) error {
					if err := callback(tx); err != nil {
						return err
					}
					return injected
				})
			}
			if _, err := closeOutboxGroup(n, group); !errors.Is(err, injected) {
				t.Fatalf("wrong failure: %v", err)
			}
			n.updateGroupState = nil
			status := "pending"
			groupStatus := "active"
			if committed {
				status = "superseded"
				groupStatus = "closed"
			}
			if outboxFor(t, n, entry.OperationID)["status"] != status {
				t.Fatal("outbox ignored actual commit")
			}
			state := apply(t, n, "group-command", map[string]any{"action": "state", "groupId": group.ID}).(map[string]any)["group"].(groupauthority.View)
			if state.Status != groupStatus {
				t.Fatal("authority and stop split")
			}
			if (len(n.Store.Reservations()) == 0) != committed {
				t.Fatal("reservation did not follow the actual commit")
			}
		})
	}
}

func TestGroupOutboxStopCommitsWithFullNormalQuota(t *testing.T) {
	n, group, entry := groupOutboxFixture(t)
	err := n.privateDatabase.Update(func(tx *groupstore.Tx) error {
		for index := 0; ; index++ {
			a, err := tx.Accounting()
			if err != nil {
				return err
			}
			remaining := a.TotalBytes - a.ReserveBytes - a.OrdinaryBytes
			if remaining <= 400 {
				return nil
			}
			if err := tx.Put("test-fill:"+operationFor(index), make([]byte, min(500000, remaining-400)), groupstore.Data); err != nil {
				return err
			}
		}
	})
	if err != nil {
		t.Fatal(err)
	}
	before, err := n.privateDatabase.Read()
	if err != nil {
		t.Fatal(err)
	}
	err = n.privateDatabase.Update(func(tx *groupstore.Tx) error { return tx.Put("test-overflow", make([]byte, 4096), groupstore.Data) })
	if !errors.Is(err, groupstore.ErrCapacity) {
		t.Fatalf("normal growth control did not fail: %v", err)
	}
	if _, err := closeOutboxGroup(n, group); err != nil {
		t.Fatal(err)
	}
	after, err := n.privateDatabase.Read()
	if err != nil {
		t.Fatal(err)
	}
	if before.Digest != after.Digest || !bytes.Equal(before.Bytes, after.Bytes) {
		t.Fatal("stop depended on a private rewrite")
	}
	apply(t, n, "lock", nil)
	apply(t, n, "unlock", map[string]any{"password": password})
	if outboxFor(t, n, entry.OperationID)["status"] != "superseded" {
		t.Fatal("quota reopened a stopped send")
	}
}

func TestGroupOutboxRejectsForgedMirrorBindingAndOrphan(t *testing.T) {
	for _, mode := range []string{"mirror", "binding", "orphan"} {
		t.Run(mode, func(t *testing.T) {
			n, group, entry := groupOutboxFixture(t)
			if mode != "mirror" {
				if _, err := closeOutboxGroup(n, group); err != nil {
					t.Fatal(err)
				}
			}
			next, err := copyPrivate(n.private, n.identity.Public.ID)
			if err != nil {
				t.Fatal(err)
			}
			current := next.Outbox[entry.OperationID]
			switch mode {
			case "mirror":
				v := true
				current.GroupStopped = &v
				next.Outbox[entry.OperationID] = current
			case "binding":
				current.GroupEpoch = core.Hash([]byte("wrong epoch"))
				next.Outbox[entry.OperationID] = current
			case "orphan":
				delete(next.Outbox, entry.OperationID)
			}
			data, err := core.Canonical(next)
			if err != nil {
				t.Fatal(err)
			}
			if _, err = n.privateDatabase.Write(data, n.privateDigest); err != nil {
				t.Fatal(err)
			}
			apply(t, n, "lock", nil)
			if _, err := n.Handle("unlock", map[string]any{"password": password}); err == nil || n.identity != nil {
				t.Fatal("corrupt stop relation opened")
			}
		})
	}
}

func TestGroupOutboxBoundedMirrorSchema(t *testing.T) {
	n, _, entry := groupOutboxFixture(t)
	records := map[string]OutboxRecord{}
	for index := 0; index < 256; index++ {
		current := entry
		current.OperationID = operationFor(index)
		current.ID = core.Hash([]byte(current.OperationID))
		stopped := index < 128
		current.GroupStopped = &stopped
		records[current.OperationID] = current
	}
	value, err := cloneValue(records)
	if err != nil {
		t.Fatal(err)
	}
	if _, err = parseOutbox(value, n.identity.Public.ID, time.Now().UnixMilli()); err != nil {
		t.Fatal(err)
	}
	record := entry
	stopped := true
	record.GroupStopped = &stopped
	for id := range record.Confirmations {
		record.Confirmations[id] = Confirmation{ReceivedAt: 10, ReadAt: 20}
	}
	if n.outboxItemLocked(record, record.Expires+1, false)["status"] != "read" {
		t.Fatal("stop erased historical delivery facts")
	}
}

func TestGroupOutboxRestrictivePrefixWithInvalidTail(t *testing.T) {
	n, group, entry := groupOutboxFixture(t)
	var anchor groups.GroupAnchor
	err := n.privateDatabase.Update(func(tx *groupstore.Tx) error {
		_, err := groupledger.Run(tx, *n.identity, func(_ *groupledger.Ledger, g *groupauthority.Registry) error {
			var err error
			anchor, err = g.Anchor(group.ID)
			return err
		})
		return err
	})
	if err != nil {
		t.Fatal(err)
	}
	closed, err := groups.CloseGroup(*n.identity, anchor, *group.Head)
	if err != nil {
		t.Fatal(err)
	}
	invalid := closed
	invalid.Signature = "invalid"
	result := apply(t, n, "group-command", map[string]any{"action": "headers", "groupId": group.ID, "headers": []groups.GroupEpoch{closed, invalid}}).(map[string]any)
	raw, err := cloneValue(result)
	if err != nil {
		t.Fatal(err)
	}
	value, _ := object(raw)
	observation, _ := object(value["observation"])
	if observation["rejected"] == nil {
		t.Fatal("invalid suffix was not reported")
	}
	if outboxFor(t, n, entry.OperationID)["status"] != "superseded" || len(n.Store.Reservations()) != 0 {
		t.Fatal("restrictive prefix failed to stop intent")
	}
}

func TestGroupOutboxStopCancelsActualLocalTransportQueue(t *testing.T) {
	for _, stop := range []bool{true, false} {
		name := "stop"
		if !stop {
			name = "negative-control-delivers"
		}
		t.Run(name, func(t *testing.T) {
			n, group, entry := groupOutboxFixture(t)
			bundle, err := n.Store.GetWithTouch(entry.ID, false)
			if err != nil {
				t.Fatal(err)
			}
			// The fixture has already stopped its sync worker. Queue an actual packet
			// before either route exists, then exercise the production group command.
			if _, err := n.Router.Broadcast(map[string]any{"type": "bundle", "bundle": bundle}, transport.Bulk, 10*time.Second, false); err != nil {
				t.Fatal(err)
			}
			if stop {
				if _, err := closeOutboxGroup(n, group); err != nil {
					t.Fatal(err)
				}
			}
			sink, err := transport.New(transport.Options{DisableRelay: true})
			if err != nil {
				t.Fatal(err)
			}
			defer sink.Close()
			address, err := sink.ListenTCP("127.0.0.1:0")
			if err != nil {
				t.Fatal(err)
			}
			disconnect, err := n.Router.ConnectTCP(address.String())
			if err != nil {
				t.Fatal(err)
			}
			defer disconnect()
			if _, err := n.Router.Broadcast(map[string]any{"witness": "reachable"}, transport.Normal, 10*time.Second, false); err != nil {
				t.Fatal(err)
			}
			count := 1
			if !stop {
				count = 2
			}
			witness, original := false, false
			for i := 0; i < count; i++ {
				ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
				got, err := sink.Next(ctx)
				cancel()
				if err != nil {
					t.Fatal(err)
				}
				witness = witness || strings.Contains(string(got.Payload), "reachable")
				original = original || strings.Contains(string(got.Payload), entry.ID)
			}
			if !witness || original == stop {
				t.Fatal("runtime stop did not control its queued packet")
			}
			ctx, cancel := context.WithTimeout(context.Background(), 300*time.Millisecond)
			defer cancel()
			if _, err := sink.Next(ctx); err == nil {
				t.Fatal("stopped packet arrived later")
			}
		})
	}
}

func TestGroupOutboxServeChecksCompletedAuthorAuthority(t *testing.T) {
	for _, mode := range []string{"active", "closed", "locked"} {
		t.Run(mode, func(t *testing.T) {
			n, group, entry := groupOutboxFixture(t)
			next, err := copyPrivate(n.private, n.identity.Public.ID)
			if err != nil {
				t.Fatal(err)
			}
			record := next.Outbox[entry.OperationID]
			for id := range record.Confirmations {
				record.Confirmations[id] = Confirmation{ReceivedAt: time.Now().UnixMilli()}
			}
			next.Outbox[entry.OperationID] = record
			if err := n.persistPrivateLocked(next); err != nil {
				t.Fatal(err)
			}
			if mode == "closed" {
				if _, err := closeOutboxGroup(n, group); err != nil {
					t.Fatal(err)
				}
			}
			witness, err := core.CreateBundle(*n.identity, "post", Content{"type": "post", "text": "request witness"}, nil, true, 60000)
			if err != nil {
				t.Fatal(err)
			}
			if _, err := n.Store.Put(witness, false); err != nil {
				t.Fatal(err)
			}
			// Reopening provides actual application receive/sync workers; the original
			// transaction-failure fixture intentionally kept those workers stopped.
			current := restartFor(t, n)
			if mode != "locked" {
				apply(t, current, "unlock", map[string]any{"password": password})
			}
			if err := current.Start(0, "127.0.0.1"); err != nil {
				t.Fatal(err)
			}
			sink, err := transport.New(transport.Options{DisableRelay: true})
			if err != nil {
				t.Fatal(err)
			}
			defer sink.Close()
			address, err := sink.ListenTCP("127.0.0.1:0")
			if err != nil {
				t.Fatal(err)
			}
			disconnect, err := current.Router.ConnectTCP(address.String())
			if err != nil {
				t.Fatal(err)
			}
			defer disconnect()
			if _, err := sink.Broadcast(map[string]any{"type": "request", "ids": []string{entry.ID, witness.Manifest.ID}}, transport.Normal, 10*time.Second, false); err != nil {
				t.Fatal(err)
			}
			gotWitness, gotPrivate, inventoryPrivate := false, false, false
			deadline := time.Now().Add(5 * time.Second)
			for time.Now().Before(deadline) {
				ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
				got, err := sink.Next(ctx)
				cancel()
				if err == nil {
					value, decodeErr := core.DecodeJSON(got.Payload, core.MaxBundleBytes+65536)
					if decodeErr != nil {
						t.Fatal(decodeErr)
					}
					m, _ := object(value)
					if m["type"] == "bundle" {
						bundle, _ := object(m["bundle"])
						manifest, _ := object(bundle["manifest"])
						gotWitness = gotWitness || text(manifest["id"]) == witness.Manifest.ID
						gotPrivate = gotPrivate || text(manifest["id"]) == entry.ID
					}
					if m["type"] == "inventory" {
						ids, _ := stringsList(m["ids"], 64, true)
						inventoryPrivate = inventoryPrivate || contains(ids, entry.ID)
					}
				}
				if gotWitness && (mode != "active" || (gotPrivate && inventoryPrivate)) {
					break
				}
			}
			if !gotWitness {
				t.Fatal("positive request witness did not arrive")
			}
			// Force one fresh inventory observation under the same application mutex.
			current.mu.Lock()
			current.syncLocked()
			current.mu.Unlock()
			deadline = time.Now().Add(400 * time.Millisecond)
			for time.Now().Before(deadline) {
				ctx, cancel := context.WithTimeout(context.Background(), 100*time.Millisecond)
				got, err := sink.Next(ctx)
				cancel()
				if err != nil {
					continue
				}
				value, _ := core.DecodeJSON(got.Payload, core.MaxBundleBytes+65536)
				m, _ := object(value)
				if m["type"] == "bundle" {
					bundle, _ := object(m["bundle"])
					manifest, _ := object(bundle["manifest"])
					gotPrivate = gotPrivate || text(manifest["id"]) == entry.ID
				}
				if m["type"] == "inventory" {
					ids, _ := stringsList(m["ids"], 64, true)
					inventoryPrivate = inventoryPrivate || contains(ids, entry.ID)
				}
			}
			if gotPrivate != (mode == "active") || inventoryPrivate != (mode == "active") {
				t.Fatal("cache serving bypassed current authority or lost positive control")
			}
			if mode != "locked" && outboxFor(t, current, entry.OperationID)["status"] != "received" {
				t.Fatal("closure erased delivery fact")
			}
		})
	}
}

File: tests/group-outbox.test.ts
import test from "node:test";
import assert from "node:assert/strict";
import { randomUUID } from "node:crypto";
import { canonical, createBundle, hash } from "../packages/core/src/index.js";
import {
  closeAnchoredGroup,
  createGroupSuccessor,
} from "../packages/groups/src/certificates.js";
import {
  validateOutbox,
  admitOutbox,
  outboxItem,
  type OutboxEntry,
} from "../apps/node/src/outbox.js";
import { password } from "./helpers.js";
import { until } from "./helpers.js";
import { Router } from "../packages/transport/src/index.js";
import { setTimeout as delay } from "node:timers/promises";

import { groupOutboxFixture as fixture } from "./fixtures/group-outbox.js";

test("missing private proof pauses a valid reserved send without losing availability, then releases its obsolete hold", (t) => {
  const f = fixture(t),
    key = `snapshot:${f.group.id}:${f.group.head.body.snapshotHash}`;
  const snapshot = f.command({
    action: "private-state",
    groupId: f.group.id,
    epochId: f.group.head.id,
  }).snapshot;
  const anchor = f.command({
    action: "proofs",
    groupId: f.group.id,
    from: 0,
  }).anchor;
  const successor = createGroupSuccessor(
    f.identity,
    anchor,
    f.group.head,
    snapshot,
    { title: "Awaiting proof", members: snapshot.members, joins: [] },
  );
  f.command({
    action: "headers",
    groupId: f.group.id,
    headers: [successor.epoch],
  });
  const state = f.node.state(),
    item = state.outbox[0];
  assert.equal(item.status, "pending");
  assert.equal(item.retained, true);
  assert.equal(item.groupAuthority?.allowed, false);
  assert.equal(item.groupAuthority?.stop, null);
  assert.equal(
    f.internals.privateState.outbox[f.entry.operationId].phase,
    "ready",
  );
  assert.equal(state.storage.reserved, 1);
  f.command({
    action: "snapshot",
    groupId: f.group.id,
    epochId: successor.epoch.id,
    snapshot: successor.snapshot,
  });
  assert.equal(f.item().groupAuthority?.allowed, true);
  // The verified cursor now points to the successor. Loss of an older private
  // snapshot is missing historical proof, not corruption of that cursor.
  f.internals.privateDatabase.transaction((tx: any) => tx.delete(key));
  const held = f.node.state();
  assert.equal(held.groupContent.held.length, 1);
  assert.equal(
    f.internals.privateState.outbox[f.entry.operationId].phase,
    "ready",
  );
  f.command({
    action: "snapshot",
    groupId: f.group.id,
    epochId: f.group.head.id,
    snapshot,
  });
  const recovered = f.node.state();
  assert.equal(recovered.outbox[0].groupAuthority?.allowed, true);
  assert.equal(recovered.groupContent.held.length, 0);
  assert.equal(
    recovered.storage.reserved,
    1,
    "pending outbox reserve survives quarantine release",
  );
});

test("runtime closure persists the minimal stop without rewriting the private document and releases only automatic protection", async (t) => {
  const f = fixture(t);
  assert.equal(f.item().status, "pending");
  assert.deepEqual(f.node.store.reservations(), [f.entry.id]);
  f.node.localAction("pin", f.entry.id, true);
  const before = f.internals.privateDatabase.read();
  f.close();
  const after = f.internals.privateDatabase.read();
  assert.deepEqual(
    after,
    before,
    "the stop must not depend on private document growth",
  );
  assert.equal(
    JSON.parse(after.bytes).outbox[f.entry.operationId].groupStopped,
    false,
  );
  assert.equal(f.item().status, "superseded");
  assert.equal(f.item().groupAuthority?.stop?.reason, "group-closed");
  assert.equal(f.node.store.isPinned(f.entry.id), true);
  assert.deepEqual(f.node.store.reservations(), []);
  const beforeStop = f.stopped();
  f.node.lock();
  f.node.unlock(password);
  assert.equal(f.item().status, "superseded");
  assert.deepEqual(f.stopped(), beforeStop);
  assert.equal(f.node.retryOutbox(f.entry.operationId).outbox.attempts, 0);
  f.node.localAction("pin", f.entry.id, false);
  assert.equal(f.node.store.isPinned(f.entry.id), false);
});

for (const committed of [false, true])
  test(`runtime group stop and closure recover together after ${committed ? "lost commit response" : "rollback"}`, (t) => {
    const f = fixture(t),
      db = f.internals.privateDatabase,
      transaction = db.transaction.bind(db);
    db.transaction = (callback: any) => {
      if (committed) {
        transaction(callback);
        throw new Error("fixture lost response");
      }
      return transaction((tx: any) => {
        callback(tx);
        throw new Error("fixture rollback");
      });
    };
    assert.throws(f.close, /fixture/);
    assert.equal(
      f.command({ action: "state", groupId: f.group.id }).group.status,
      committed ? "closed" : "active",
    );
    assert.equal(!!f.stopped(), committed);
    assert.equal(f.item().status, committed ? "superseded" : "pending");
    assert.equal(f.node.store.reservations().includes(f.entry.id), !committed);
  });

test("full normal database quota still commits a group stop while the ordinary growth control fails", (t) => {
  const f = fixture(t),
    db = f.internals.privateDatabase;
  db.transaction((tx: any) => {
    let index = 0;
    while (true) {
      const accounting = tx.accounting();
      const remaining =
        accounting.totalBytes -
        accounting.reserveBytes -
        accounting.ordinaryBytes;
      if (remaining <= 400) break;
      tx.put(
        "test-fill:" + index++,
        Buffer.alloc(Math.min(500_000, remaining - 400)),
      );
    }
  });
  const before = db.read();
  assert.throws(
    () =>
      db.transaction((tx: any) => tx.put("test-overflow", Buffer.alloc(4096))),
    /limite|espaço|capacidade/i,
  );
  f.close();
  assert.deepEqual(db.read(), before);
  assert.equal(f.item().status, "superseded");
  f.node.lock();
  f.node.unlock(password);
  assert.equal(f.item().status, "superseded");
});

test("forged stop mirror, mismatched binding and orphan stop all fail authenticated reopening", async (t) => {
  for (const mode of ["mirror", "binding", "orphan"] as const)
    await t.test(mode, (t) => {
      const f = fixture(t);
      if (mode !== "mirror") f.close();
      const document = f.internals.privateDatabase.read();
      const value = JSON.parse(document.bytes);
      const entry = value.outbox[f.entry.operationId];
      if (mode === "mirror") entry.groupStopped = true;
      if (mode === "binding") entry.groupEpoch = "a".repeat(64);
      if (mode === "orphan") delete value.outbox[f.entry.operationId];
      f.internals.privateDatabase.write(
        Buffer.from(canonical(value)),
        document.digest,
      );
      f.node.lock();
      assert.throws(
        () => f.node.unlock(password),
        /Paragem|paragem|intenção|Admissão/,
      );
      assert.equal(f.node.identity, undefined);
    });
});

test("bounded stopped entries coexist with 128 new pending operations; half bindings and extra fields fail", (t) => {
  const f = fixture(t),
    old: Record<string, OutboxEntry> = {};
  for (let n = 0; n < 128; n++) {
    const operationId = randomUUID();
    old[operationId] = {
      ...structuredClone(f.entry),
      operationId,
      id: hash("old" + n),
      groupStopped: true,
    };
  }
  let next = old;
  for (let n = 0; n < 128; n++) {
    const operationId = randomUUID();
    next = admitOutbox(
      next,
      {
        ...structuredClone(f.entry),
        operationId,
        id: hash("new" + n),
        groupStopped: false,
      },
      Date.now(),
    );
  }
  assert.equal(
    Object.keys(validateOutbox(next, f.identity.public.id)).length,
    256,
  );
  assert.throws(
    () =>
      admitOutbox(next, { ...f.entry, operationId: randomUUID() }, Date.now()),
    /cheia/,
  );
  for (const patch of [
    { groupEpoch: undefined },
    { groupStopped: undefined },
    { conversation: "dm:" + hash("x") },
    { extra: true },
  ]) {
    const entry = JSON.parse(JSON.stringify({ ...f.entry, ...patch }));
    assert.throws(() =>
      validateOutbox({ [entry.operationId]: entry }, f.identity.public.id),
    );
  }
  const completed = {
    ...f.entry,
    groupStopped: true,
    confirmations: { [f.reader.public.id]: { receivedAt: 10, readAt: 20 } },
  };
  assert.equal(outboxItem(completed, completed.expires + 1, []).status, "read");
  assert.equal(
    outboxItem(completed, completed.expires + 1, []).contentExpired,
    true,
  );
});

test("a restrictive prefix with an invalid tail stops the runtime intent in its accepted transaction", (t) => {
  const f = fixture(t);
  const anchor = f.command({
    action: "proofs",
    groupId: f.group.id,
    from: 0,
  }).anchor;
  const closed = closeAnchoredGroup(f.identity, anchor, f.group.head);
  const response = f.command({
    action: "headers",
    groupId: f.group.id,
    headers: [
      closed,
      { ...closed, signature: Buffer.alloc(64).toString("base64") },
    ],
  });
  assert.ok(response.observation.rejected);
  assert.equal(f.item().status, "superseded");
  assert.equal(f.stopped().epochId, f.entry.groupEpoch);
  assert.deepEqual(f.node.store.reservations(), []);
});

test("stop retirement and removal of the corresponding intent share a private commit", (t) => {
  const f = fixture(t);
  f.close();
  const db = f.internals.privateDatabase,
    transaction = db.transaction.bind(db);
  const next = structuredClone(f.internals.privateState);
  delete next.outbox[f.entry.operationId];
  db.transaction = (callback: any) =>
    transaction((tx: any) => {
      callback(tx);
      throw new Error("fixture retirement rollback");
    });
  assert.throws(() => f.internals.persistPrivate(next), /fixture retirement/);
  assert.equal(f.item().status, "superseded");
  assert.ok(f.stopped());
  f.internals.persistPrivate(next);
  assert.equal(f.stopped(), null);
  f.node.lock();
  f.node.unlock(password);
  assert.equal(f.node.state().outbox.length, 0);
});

for (const stop of [true, false])
  test(`runtime ${stop ? "stop cancels" : "negative control delivers"} its existing transport queue`, async (t) => {
    const f = fixture(t),
      sink = new Router({ relay: false });
    t.after(() => sink.stop());
    const received: any[] = [];
    sink.on("payload", (p) => received.push(p));
    f.node.router.broadcast({ type: "bundle", bundle: f.bundle }, "bulk");
    if (stop) f.close();
    f.node.router.connectTcp("127.0.0.1", await sink.listen());
    f.node.router.broadcast({ witness: "reachable" });
    await until(
      async () => received,
      (list) => list.some((p) => p.witness === "reachable"),
    );
    if (!stop)
      await until(
        async () => received,
        (list) => list.some((p) => p.bundle?.manifest.id === f.entry.id),
      );
    await delay(300);
    assert.equal(
      received.some((p) => p.bundle?.manifest.id === f.entry.id),
      !stop,
    );
    assert.equal(
      f.item().attempts,
      0,
      "fixture did not invent an outbox transmission attempt",
    );
  });

for (const mode of ["active", "closed", "locked"] as const)
  test(`real TCP cache request checks ${mode} author authority even after delivery completed`, async (t) => {
    const f = fixture(t),
      sink = new Router({ relay: false });
    t.after(() => sink.stop());
    const next = structuredClone(f.internals.privateState);
    next.outbox[f.entry.operationId].confirmations[f.reader.public.id] = {
      receivedAt: Date.now(),
    };
    f.internals.persistPrivate(next);
    assert.equal(f.item().status, "received");
    if (mode === "closed") f.close();
    // Stored witness has no initial outgoing packet. Its later receipt proves
    // the request reached the real application serving branch on this socket.
    const witness = createBundle(
      f.identity,
      "post",
      { type: "post", text: "request witness" },
      "public",
      60000,
    );
    f.node.store.put(witness);
    if (mode === "locked") f.node.lock();
    const received: any[] = [];
    sink.on("payload", (p) => received.push(p));
    f.node.router.connectTcp("127.0.0.1", await sink.listen());
    await until(
      async () => sink.peers,
      (peers) => peers.some((p) => p.connected),
    );
    sink.broadcast({ type: "request", ids: [f.entry.id, witness.manifest.id] });
    await until(
      async () => received,
      (list) => list.some((p) => p.bundle?.manifest.id === witness.manifest.id),
    );
    if (mode === "active")
      await until(
        async () => received,
        (list) => list.some((p) => p.bundle?.manifest.id === f.entry.id),
      );
    f.node.sync();
    await delay(350);
    assert.equal(
      received.some((p) => p.bundle?.manifest.id === f.entry.id),
      mode === "active",
    );
    assert.equal(
      received.some(
        (p) => p.type === "inventory" && p.ids.includes(f.entry.id),
      ),
      mode === "active",
    );
    if (mode !== "locked")
      assert.equal(
        f.item().status,
        "received",
        "closure must preserve the prior delivery fact",
      );
  });
